File: runtime/java/src/main/java/dev/argon/verilization/java_runtime/Option.java
package dev.argon.verilization.java_runtime;

import java.util.Optional;
import java.io.IOException;

public final class Option {
    private Option() {}

    public static <A, B> Converter<Optional<A>, Optional<B>> converter(Converter<A, B> elementConverter) {
        return prev -> prev.map(elementConverter::convert);
    }

    public static <T> Codec<Optional<T>> codec(Codec<T> elementCodec) {
        return new Codec<Optional<T>>() {
            @Override
            public Optional<T> read(FormatReader reader) throws IOException {
                boolean present = reader.readByte() != 0;
                if(present) {
                    return Optional.of(elementCodec.read(reader));
                }
                else {
                    return Optional.empty();
                }
            }
    
            @Override
            public void write(FormatWriter writer, Optional<T> value) throws IOException {
                if(value.isPresent()) {
                    writer.writeByte((byte)1);
                    elementCodec.write(writer, value.get());
                }
                else {
                    writer.writeByte((byte)0);
                }
            }
    
        };
    }
}

File: runtime/scala/shared/src/main/scala/dev/argon/verilization/scala_runtime/Option.scala
package dev.argon.verilization.scala_runtime

import zio.{ZIO, IO}
import scala.{Option => SOption}

object Option {
    def fromCaseSome[A](value: A): Option[A] = Some(value)
    def fromCaseNone[A](): Option[A] = None

    def converter[A, B](elementConverter: Converter[A, B]): Converter[SOption[A], SOption[B]] = elementConverter match {
        case _: IdentityConverter[_] => new IdentityConverter[SOption[A]]
        case _ => new Converter[SOption[A], SOption[B]] {
            override def convert(prev: SOption[A]): SOption[B] = prev.map(elementConverter.convert)
        }
    }

    def codec[A](elementCodec: Codec[A]): Codec[SOption[A]] = new Codec[SOption[A]] {
        override def read[R, E](reader: FormatReader[R, E]): ZIO[R, E, SOption[A]] =
            U8.codec.read(reader).flatMap { present =>
                if(present != 0) {
                    elementCodec.read(reader).map(Some.apply)
                }
                else {
                    IO.none
                }
            }

        override def write[R, E](writer: FormatWriter[R, E], value: SOption[A]): ZIO[R, E, Unit] =
            U8.codec.write(writer, if(value.isDefined) 1 else 0) *> ZIO.foreachDiscard(value) { elem => elementCodec.write(writer, elem) }
    }
}
